src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -926,77 +986,67 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
+        private async ValueTask SendStreamDataAsync(int streamId, ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
         {
             ReadOnlyMemory<byte> remaining = buffer;
 
             while (remaining.Length > 0)
             {
                 int frameSize = Math.Min(remaining.Length, FrameHeader.MaxLength);
 
-                frameSize = await _connectionWindow.RequestCreditAsync(frameSize).ConfigureAwait(false);
+                // Once credit had been granted, we want to actually consume those bytes.
+                frameSize = await _connectionWindow.RequestCreditAsync(frameSize, cancellationToken).ConfigureAwait(false);
 
                 ReadOnlyMemory<byte> current;
                 (current, remaining) = SplitBuffer(remaining, frameSize);
 
-                await AcquireWriteLockAsync().ConfigureAwait(false);
+                // It's possible that a cancellation will occur while we wait for the write lock. In that case, we need to
+                // return the credit that we have acquired and don't plan to use.
                 try
                 {
-                    _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + current.Length);
-                    WriteFrameHeader(new FrameHeader(current.Length, FrameType.Data, FrameFlags.None, streamId));
-                    current.CopyTo(_outgoingBuffer.AvailableMemory);
-                    _outgoingBuffer.Commit(current.Length);
-
-                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                    await StartWriteAsync(FrameHeader.Size + current.Length, cancellationToken).ConfigureAwait(false);
                 }
-                finally
+                catch (OperationCanceledException)
                 {
-                    ReleaseWriteLock();
+                    _connectionWindow.AdjustCredit(frameSize);
+                    throw;
                 }
+
+                WriteFrameHeader(new FrameHeader(current.Length, FrameType.Data, FrameFlags.None, streamId));
+                current.CopyTo(_outgoingBuffer.AvailableMemory);
+                _outgoingBuffer.Commit(current.Length);
+
+                await FinishWriteAsync(false, cancellationToken).ConfigureAwait(false);
             }
         }
 
         private async ValueTask SendEndStreamAsync(int streamId)
         {
-            await AcquireWriteLockAsync().ConfigureAwait(false);
-            try
-            {
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size);
-                WriteFrameHeader(new FrameHeader(0, FrameType.Data, FrameFlags.EndStream, streamId));
+            await StartWriteAsync(FrameHeader.Size).ConfigureAwait(false);
 
-                await FlushOutgoingBytesAsync().ConfigureAwait(false);
-            }
-            finally
-            {
-                ReleaseWriteLock();
-            }
+            WriteFrameHeader(new FrameHeader(0, FrameType.Data, FrameFlags.EndStream, streamId));
```

Comment:
Is there any situation in which WriteFrameHeader could throw?  If it could, it seems like we'd end up deadlocking because we'd never release the lock?

Reply:
We have some Debug.Asserts that could fail in WriteFrameHeader, but I don't think that we can throw otherwise.